File: cpp/benchmarks/benchmark_roundtrip.cpp
#include <zerobuffer/zerobuffer.h>
#include <zerobuffer/platform.h>
#include <iostream>
#include <iomanip>
#include <chrono>
#include <vector>
#include <algorithm>
#include <numeric>
#include <thread>
#include <atomic>
#include <csignal>
#include <unistd.h>
#include <sys/wait.h>
#include <cstring>

using namespace zerobuffer;
using namespace std::chrono;

// Frame structure matching C# implementation
#pragma pack(push, 1)
struct TimestampedFrame {
    int64_t timestamp;    // 8 bytes - high resolution timestamp
    int32_t frame_id;     // 4 bytes
    int32_t padding;      // 4 bytes - align to 16 bytes
    // Followed by YUV420 data
};
#pragma pack(pop)

// Constants
constexpr size_t YUV420_FRAME_SIZE = 1920 * 1080 * 3 / 2;  // 3,110,400 bytes
constexpr size_t HEADER_SIZE = sizeof(TimestampedFrame);
constexpr size_t TOTAL_FRAME_SIZE = HEADER_SIZE + YUV420_FRAME_SIZE;
constexpr size_t BUFFER_SIZE = 256 * 1024 * 1024;  // 256 MB
constexpr int WARMUP_FRAMES = 100;

// Test configurations
struct TestConfig {
    int target_fps;
    int frame_count;
};

const std::vector<TestConfig> TEST_CONFIGS = {
    {30, 100},   // Further reduced for testing
    {60, 100},   // Further reduced for testing
    //{120, 1000},
    //{240, 1000},
    //{500, 1000},
    //{1000, 1000}
};

// Forward declarations
double ticks_to_microseconds(int64_t ticks);

class LatencyBenchmark {
private:
    std::vector<double> latencies_us_;
    high_resolution_clock::time_point start_time_;
    
public:
    void start() {
        latencies_us_.clear();
        start_time_ = high_resolution_clock::now();
    }
    
    void record_latency(int64_t send_ticks, int64_t receive_ticks) {
        // Convert ticks to microseconds
        double latency_us = ticks_to_microseconds(receive_ticks - send_ticks);
        if (latency_us > 0) {  // Only record positive latencies
            latencies_us_.push_back(latency_us);
        }
    }
    
    void print_results(int target_fps, int frames_sent, int frames_received) {
        if (latencies_us_.empty()) {
            std::cout << "  No data collected!" << std::endl;
            return;
        }
        
        // Sort for percentile calculations
        std::sort(latencies_us_.begin(), latencies_us_.end());
        
        double min_us = latencies_us_.front();
        double max_us = latencies_us_.back();
        double avg_us = std::accumulate(latencies_us_.begin(), latencies_us_.end(), 0.0) 
                        / latencies_us_.size();
        
        size_t p50_idx = latencies_us_.size() * 50 / 100;
        size_t p90_idx = latencies_us_.size() * 90 / 100;
        size_t p99_idx = latencies_us_.size() * 99 / 100;
        
        double p50_us = latencies_us_[p50_idx];
        double p90_us = latencies_us_[p90_idx];
        double p99_us = latencies_us_[p99_idx];
        
        std::cout << "  Frames sent: " << frames_sent 
                  << ", received: " << frames_received << std::endl;
        std::cout << "  Round-trip latency (microseconds):" << std::endl;
        std::cout << "    Min:    " << std::fixed << std::setprecision(0) 
                  << min_us << " μs" << std::endl;
        std::cout << "    Avg:    " << avg_us << " μs" << std::endl;
        std::cout << "    P50:    " << p50_us << " μs" << std::endl;
        std::cout << "    P90:    " << p90_us << " μs" << std::endl;
        std::cout << "    P99:    " << p99_us << " μs" << std::endl;
        std::cout << "    Max:    " << max_us << " μs" << std::endl;
    }
};

// Get high-resolution timestamp in ticks (platform specific)
int64_t get_timestamp_ticks() {
    auto now = high_resolution_clock::now();
    return now.time_since_epoch().count();
}

// Convert ticks to microseconds
double ticks_to_microseconds(int64_t ticks) {
    // Get the tick frequency
    using period = high_resolution_clock::period;
    double ticks_per_second = static_cast<double>(period::den) / period::num;
    return (ticks * 1000000.0) / ticks_per_second;
}

// High-precision periodic timer
class PeriodicTimer {
private:
    high_resolution_clock::time_point next_tick_;
    std::chrono::microseconds interval_;
    
public:
    explicit PeriodicTimer(std::chrono::microseconds interval) 
        : interval_(interval) {
        // Minimum 1ms interval like C#
        if (interval_ < microseconds(1000)) {
            interval_ = microseconds(1000);
        }
        next_tick_ = high_resolution_clock::now() + interval_;
    }
    
    void wait_for_next_tick() {
        auto now = high_resolution_clock::now();
        
        // If we're running late, just update next tick
        if (now >= next_tick_) {
            // Calculate how many intervals we missed
            auto elapsed = duration_cast<microseconds>(now - next_tick_);
            auto missed_intervals = elapsed.count() / interval_.count() + 1;
            next_tick_ += interval_ * missed_intervals;
            return;
        }
        
        // Busy wait for the last microsecond for precision
        auto wait_until = next_tick_ - microseconds(1000);
        if (now < wait_until) {
            std::this_thread::sleep_until(wait_until);
        }
        
        // Busy wait for final precision
        while (high_resolution_clock::now() < next_tick_) {
            std::this_thread::yield();
        }
        
        next_tick_ += interval_;
    }
};

void run_benchmark_at_fps(int target_fps, int frame_count) {
    std::cout << "\n--- Testing at " << target_fps << " FPS ---" << std::endl;
    
    // Create buffers outside try block so they stay alive
    std::unique_ptr<Reader> reader;
    std::unique_ptr<Writer> writer;
    
    try {
        // Create Buffer B as Reader (output buffer)
        BufferConfig config(0, BUFFER_SIZE);
        reader = std::make_unique<Reader>("buffer-b", config);
        std::cout << "  Created buffer-b as Reader" << std::endl;
        
        // Give relay more time to initialize and connect
        std::this_thread::sleep_for(milliseconds(1000));
        
        // Connect to Buffer A as Writer (input buffer)
        writer = std::make_unique<Writer>("buffer-a");
        std::cout << "  Connected to buffer-a as Writer" << std::endl;
        
        // Prepare frame data
        std::vector<uint8_t> frame_data(TOTAL_FRAME_SIZE);
        
        // Fill YUV data with pattern (not important for latency test)
        for (size_t i = HEADER_SIZE; i < TOTAL_FRAME_SIZE; ++i) {
            frame_data[i] = static_cast<uint8_t>(i % 256);
        }
        
        // Calculate frame interval
        auto frame_interval = microseconds(1000000 / target_fps);
        PeriodicTimer timer(frame_interval);
        
        // Start a receiver thread for warmup too
        std::atomic<int> warmup_received(0);
        std::atomic<bool> warmup_done(false);
        std::thread warmup_receiver([&]() {
            std::cout << "  [DEBUG] Warmup receiver thread started" << std::endl;
            while (!warmup_done) {
                try {
                    std::cout << "  [DEBUG] Waiting for warmup frame..." << std::endl;
                    Frame frame = reader->read_frame();
                    if (frame.valid()) {
                        if (frame.size() >= HEADER_SIZE) {
                            const TimestampedFrame* header = 
                                reinterpret_cast<const TimestampedFrame*>(frame.data());
                            std::cout << "  [DEBUG] Received frame id=" << header->frame_id << std::endl;
                            if (header->frame_id < 0) {  // Warmup frame
                                warmup_received++;
                            }
                        }
                        reader->release_frame(frame);
                    }
                } catch (const std::exception& e) {
                    std::cout << "  [DEBUG] Warmup receiver error: " << e.what() << std::endl;
                    break;
                }
            }
            std::cout << "  [DEBUG] Warmup receiver thread exiting" << std::endl;
        });
        
        // Warmup
        std::cout << "  Warming up... " << std::flush;
        for (int i = 0; i < WARMUP_FRAMES; ++i) {
            // Get timestamp right before sending
            int64_t send_ticks = get_timestamp_ticks();
            
            TimestampedFrame* header = reinterpret_cast<TimestampedFrame*>(frame_data.data());
            header->timestamp = send_ticks;
            header->frame_id = -(i+1);  // Negative for warmup frames
            
            std::cout << "  [DEBUG] Sending warmup frame " << i+1 << "/" << WARMUP_FRAMES 
                      << " (id=" << header->frame_id << ")" << std::endl;
            
            // Send using zero-copy
            uint64_t sequence;
            void* buffer = writer->get_frame_buffer(TOTAL_FRAME_SIZE, sequence);
            std::memcpy(buffer, frame_data.data(), TOTAL_FRAME_SIZE);
            writer->commit_frame();
            
            timer.wait_for_next_tick();
        }
        
        // Wait for warmup frames to be received
        std::cout << "  [DEBUG] Waiting for warmup frames to be received..." << std::endl;
        auto warmup_timeout = high_resolution_clock::now() + seconds(5);
        while (warmup_received < WARMUP_FRAMES && high_resolution_clock::now() < warmup_timeout) {
            std::cout << "  [DEBUG] Warmup received: " << warmup_received << "/" << WARMUP_FRAMES << std::endl;
            std::this_thread::sleep_for(milliseconds(100));
        }
        
        std::cout << "  [DEBUG] Setting warmup_done flag" << std::endl;
        warmup_done = true;
        std::cout << "  [DEBUG] Joining warmup receiver thread..." << std::endl;
        warmup_receiver.join();
        std::cout << "done (received " << warmup_received << " frames)" << std::endl;
        
        // Measurement
        LatencyBenchmark benchmark;
        benchmark.start();
        
        int frames_to_send = frame_count;
        int frames_sent = 0;
        int frames_received = 0;
        
        std::cout << "  Measuring " << frames_to_send << " frames... " << std::flush;
        
        // Main test loop - similar to C# pattern
        for (int i = 0; i < frames_to_send; ++i) {
            // Send frame with timestamp
            int64_t send_ticks = get_timestamp_ticks();
            
            TimestampedFrame* header = reinterpret_cast<TimestampedFrame*>(frame_data.data());
            header->timestamp = send_ticks;
            header->frame_id = i;
            
            // Send frame using zero-copy
            uint64_t sequence;
            void* buffer = writer->get_frame_buffer(TOTAL_FRAME_SIZE, sequence);
            std::memcpy(buffer, frame_data.data(), TOTAL_FRAME_SIZE);
            writer->commit_frame();
            frames_sent++;
            
            // Note: C++ doesn't support non-blocking reads, so we'll need
            // to use a different pattern - collect responses in a separate thread
            
            timer.wait_for_next_tick();
        }
        
        // Start receiver thread
        std::atomic<bool> receiver_done(false);
        std::thread receiver([&]() {
            while (!receiver_done) {
                try {
                    Frame frame = reader->read_frame();
                    if (!frame.valid()) {
                        continue;
                    }
                    
                    // Get receive timestamp immediately
                    int64_t receive_ticks = get_timestamp_ticks();
                    
                    // Extract the timestamp from frame data
                    if (frame.size() >= HEADER_SIZE) {
                        const TimestampedFrame* header = 
                            reinterpret_cast<const TimestampedFrame*>(frame.data());
                        
                        // Process all frames (including warmup)
                        if (header->frame_id >= 0) {
                            benchmark.record_latency(header->timestamp, receive_ticks);
                            frames_received++;
                            
                            if (frames_received >= frames_to_send) {
                                receiver_done = true;
                            }
                        }
                    }
                    
                    reader->release_frame(frame);
                    
                } catch (const WriterDeadException&) {
                    break;
                } catch (const std::exception& e) {
                    std::cerr << "Receiver error: " << e.what() << std::endl;
                    break;
                }
            }
        });
        
        // Send frames
        for (int i = 0; i < frames_to_send; ++i) {
            // Send frame with timestamp
            int64_t send_ticks = get_timestamp_ticks();
            
            TimestampedFrame* header = reinterpret_cast<TimestampedFrame*>(frame_data.data());
            header->timestamp = send_ticks;
            header->frame_id = i;
            
            // Send frame using zero-copy
            uint64_t sequence;
            void* buffer = writer->get_frame_buffer(TOTAL_FRAME_SIZE, sequence);
            std::memcpy(buffer, frame_data.data(), TOTAL_FRAME_SIZE);
            writer->commit_frame();
            frames_sent++;
            
            timer.wait_for_next_tick();
        }
        
        // Wait for receiver to finish (with timeout)
        auto timeout = high_resolution_clock::now() + seconds(5);
        while (!receiver_done && high_resolution_clock::now() < timeout) {
            std::this_thread::sleep_for(milliseconds(10));
        }
        
        receiver_done = true;
        receiver.join();
        
        std::cout << "done" << std::endl;
        
        // Print results
        benchmark.print_results(target_fps, frames_sent, frames_received);
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }
}

pid_t start_relay_process() {
    pid_t relay_pid = fork();
    if (relay_pid == 0) {
        // Child process - execute relay
        execl("./relay_process", "relay_process", nullptr);
        std::cerr << "Failed to execute relay process" << std::endl;
        exit(1);
    }
    return relay_pid;
}

void stop_relay_process(pid_t relay_pid) {
    kill(relay_pid, SIGTERM);
    int status;
    waitpid(relay_pid, &status, 0);
    
    // Clean up shared memory resources
    // Note: On Linux, semaphore names don't have a leading slash in the filesystem
    try {
        SharedMemory::remove("buffer-a");
        SharedMemory::remove("buffer-b");
        Semaphore::remove("sem-w-buffer-a");
        Semaphore::remove("sem-r-buffer-a");
        Semaphore::remove("sem-w-buffer-b");
        Semaphore::remove("sem-r-buffer-b");
    } catch (...) {
        // Ignore cleanup errors
    }
}

int main() {
    std::cout << "ZeroBuffer Cross-Process Round-Trip Latency Benchmark" << std::endl;
    std::cout << "=====================================================" << std::endl;
    std::cout << "Frame size: " << TOTAL_FRAME_SIZE << " bytes "
              << "(YUV420 1920x1080 + 16-byte header)" << std::endl;
    std::cout << "Buffer size: " << (BUFFER_SIZE / 1024 / 1024) << " MB" << std::endl;
    
    // Initial cleanup to ensure we start fresh
    try {
        SharedMemory::remove("buffer-a");
        SharedMemory::remove("buffer-b");
        Semaphore::remove("sem-w-buffer-a");
        Semaphore::remove("sem-r-buffer-a");
        Semaphore::remove("sem-w-buffer-b");
        Semaphore::remove("sem-r-buffer-b");
    } catch (...) {
        // Ignore cleanup errors
    }
    
    // Run benchmarks at different FPS levels
    for (const auto& config : TEST_CONFIGS) {
        // Start fresh relay process for each test
        pid_t relay_pid = start_relay_process();
        
        // Give relay time to initialize
        std::this_thread::sleep_for(seconds(2));
        
        run_benchmark_at_fps(config.target_fps, config.frame_count);
        
        // Stop relay process
        stop_relay_process(relay_pid);
        
        // Brief pause between tests
        std::this_thread::sleep_for(seconds(1));
    }
    
    std::cout << "\nBenchmark complete!" << std::endl;
    
    return 0;
}

File: python/benchmark_roundtrip.py
#!/usr/bin/env python3
"""
ZeroBuffer Cross-Process Round-Trip Latency Benchmark

Measures round-trip latency for video frames through a relay process.
Similar to the C# benchmark implementation.
"""

import os
import sys
import time
import struct
import random
import asyncio
import statistics
import subprocess
from typing import List, Tuple
from datetime import datetime

from zerobuffer import Reader, Writer, BufferConfig
from zerobuffer.exceptions import ZeroBufferException


# Frame structure constants
HEADER_SIZE = 16  # 8 bytes timestamp + 4 bytes frame_id + 4 bytes padding
FULLHD_1080P_SIZE = 1920 * 1080 * 3 // 2  # YUV420: 3,110,400 bytes
TOTAL_FRAME_SIZE = HEADER_SIZE + FULLHD_1080P_SIZE

# Test parameters
WARMUP_FRAMES = 100
TEST_FRAMES_PER_FPS = 1000


def get_timestamp_ns() -> int:
    """Get high-resolution timestamp in nanoseconds"""
    return time.perf_counter_ns()


def pack_frame_header(timestamp: int, frame_id: int) -> bytes:
    """Pack timestamp and frame ID into header"""
    # Q = unsigned long long (8 bytes), I = unsigned int (4 bytes)
    return struct.pack('<QII', timestamp, frame_id, 0)  # 0 is padding


def unpack_frame_header(data: bytes) -> Tuple[int, int]:
    """Unpack timestamp and frame ID from header"""
    timestamp, frame_id, _ = struct.unpack('<QII', data[:HEADER_SIZE])
    return timestamp, frame_id


async def start_relay_process(input_buffer: str, output_buffer: str) -> subprocess.Popen:
    """Start the relay helper process"""
    relay_script = os.path.join(os.path.dirname(__file__), 'zerobuffer_relay.py')
    
    cmd = [sys.executable, relay_script, input_buffer, output_buffer]
    
    process = subprocess.Popen(
        cmd,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True
    )
    
    # Wait for relay to initialize
    await asyncio.sleep(1.0)
    
    if process.poll() is not None:
        stderr = process.stderr.read() if process.stderr else ""
        raise RuntimeError(f"Relay process failed to start: {stderr}")
    
    return process


async def measure_latency(target_fps: int) -> None:
    """Measure round-trip latency at specified FPS"""
    test_id = f"{os.getpid()}_{int(time.time())}"
    buffer_to_relay = f"bench_to_relay_{test_id}"
    buffer_from_relay = f"bench_from_relay_{test_id}"
    
    # Start relay process
    relay_process = await start_relay_process(buffer_to_relay, buffer_from_relay)
    
    try:
        await run_benchmark(buffer_to_relay, buffer_from_relay, target_fps)
    finally:
        # Stop relay process
        relay_process.terminate()
        try:
            relay_process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            relay_process.kill()
            relay_process.wait()


async def run_benchmark(buffer_to_relay: str, buffer_from_relay: str, target_fps: int) -> None:
    """Run the actual benchmark"""
    config = BufferConfig(metadata_size=4096, payload_size=256 * 1024 * 1024)  # 256MB
    latencies = []
    
    # Prepare frame data
    frame_data = bytearray(TOTAL_FRAME_SIZE)
    # Random video data after header
    frame_data[HEADER_SIZE:] = random.randbytes(FULLHD_1080P_SIZE)
    
    # Create our output buffer as reader (relay will connect as writer)
    reader_from_relay = Reader(buffer_from_relay, config)
    
    # Give relay time to connect to our output buffer
    await asyncio.sleep(0.5)
    
    # Connect to relay's input buffer as writer
    writer_to_relay = None
    for retry in range(20):  # 2 second timeout
        try:
            writer_to_relay = Writer(buffer_to_relay)
            break
        except ZeroBufferException:
            await asyncio.sleep(0.1)
    
    if writer_to_relay is None:
        raise RuntimeError(f"Failed to connect to relay input buffer: {buffer_to_relay}")
    
    try:
        frame_interval = 1.0 / target_fps
        
        # Warmup
        print("  Warming up...", end='', flush=True)
        for i in range(WARMUP_FRAMES):
            # Pack header
            timestamp = get_timestamp_ns()
            frame_data[:HEADER_SIZE] = pack_frame_header(timestamp, i)
            
            writer_to_relay.write_frame(bytes(frame_data))
            
            # Try to read response
            response = reader_from_relay.read_frame(timeout=0.01)
            if response:
                reader_from_relay.release_frame(response)
            
            await asyncio.sleep(frame_interval)
        
        print(" done")
        
        # Clear any pending frames
        while True:
            frame = reader_from_relay.read_frame(timeout=0)
            if not frame:
                break
            reader_from_relay.release_frame(frame)
        
        # Main test
        print(f"  Measuring {TEST_FRAMES_PER_FPS} frames...", end='', flush=True)
        frames_sent = 0
        frames_received = 0
        
        # Send frames at target FPS
        start_time = time.perf_counter()
        next_frame_time = start_time
        
        for i in range(TEST_FRAMES_PER_FPS):
            # Send frame with timestamp
            send_timestamp = get_timestamp_ns()
            frame_data[:HEADER_SIZE] = pack_frame_header(send_timestamp, i)
            writer_to_relay.write_frame(bytes(frame_data))
            frames_sent += 1
            
            # Try to read response immediately
            response = reader_from_relay.read_frame(timeout=0.05)
            if response and len(response.data) >= HEADER_SIZE:
                receive_timestamp = get_timestamp_ns()
                
                # Extract original timestamp
                original_timestamp, frame_id = unpack_frame_header(bytes(response.data))
                
                # Calculate round-trip latency in microseconds
                latency_ns = receive_timestamp - original_timestamp
                latency_us = latency_ns / 1000.0
                latencies.append(latency_us)
                frames_received += 1
                
                reader_from_relay.release_frame(response)
            
            # Wait for next frame time
            next_frame_time += frame_interval
            sleep_time = next_frame_time - time.perf_counter()
            if sleep_time > 0:
                await asyncio.sleep(sleep_time)
        
        # Read any remaining responses
        await asyncio.sleep(0.1)
        while True:
            response = reader_from_relay.read_frame(timeout=0)
            if not response:
                break
            
            if len(response.data) >= HEADER_SIZE:
                receive_timestamp = get_timestamp_ns()
                original_timestamp, frame_id = unpack_frame_header(bytes(response.data))
                latency_ns = receive_timestamp - original_timestamp
                latency_us = latency_ns / 1000.0
                latencies.append(latency_us)
                frames_received += 1
            
            reader_from_relay.release_frame(response)
        
        print(" done")
        
        # Print results
        print(f"  Frames sent: {frames_sent}, received: {frames_received}")
        
        if latencies:
            latencies.sort()
            min_lat = min(latencies)
            max_lat = max(latencies)
            avg_lat = statistics.mean(latencies)
            p50_lat = statistics.quantiles(latencies, n=100)[49]  # 50th percentile
            p90_lat = statistics.quantiles(latencies, n=100)[89]  # 90th percentile
            p99_lat = statistics.quantiles(latencies, n=100)[98]  # 99th percentile
            
            print("  Round-trip latency (microseconds):")
            print(f"    Min:  {min_lat:8.0f} μs")
            print(f"    Avg:  {avg_lat:8.0f} μs")
            print(f"    P50:  {p50_lat:8.0f} μs")
            print(f"    P90:  {p90_lat:8.0f} μs")
            print(f"    P99:  {p99_lat:8.0f} μs")
            print(f"    Max:  {max_lat:8.0f} μs")
        else:
            print("  ERROR: No latency measurements collected!")
            
    finally:
        writer_to_relay.close()
        reader_from_relay.close()


async def run_all_tests() -> None:
    """Run tests at various FPS levels"""
    print("ZeroBuffer Cross-Process Round-Trip Latency Benchmark")
    print("=====================================================")
    print()
    print(f"Frame size: YUV420 Full HD + timestamp header ({TOTAL_FRAME_SIZE:,} bytes)")
    print(f"Warmup frames: {WARMUP_FRAMES}")
    print(f"Test frames per FPS: {TEST_FRAMES_PER_FPS}")
    print()
    
    # Test at various FPS levels
    fps_targets = [30, 60, 120, 240, 500, 1000]
    
    for fps in fps_targets:
        print(f"--- Testing at {fps} FPS ---")
        await measure_latency(fps)
        print()
        
        # Small delay between tests
        await asyncio.sleep(1.0)


async def main():
    """Main entry point"""
    try:
        await run_all_tests()
    except KeyboardInterrupt:
        print("\nBenchmark interrupted by user")
    except Exception as e:
        print(f"\nError: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    asyncio.run(main())